src/Request/RequestHandler.php

```diff
@@ -317,4 +352,96 @@ public function request($httpMethod, $service, $parameters = [ ])
         );
     }
+
+    /**
+     * Generates oauth signature
+     */
+    public function generateOauthSignature(string $method, string $path, array $parameters = [])
+    {
+        switch ($this->signatureMethod) {
+            case 'RSA-SHA1':
+                return $this->generateRSASHA1Signature($method, $path, $parameters);
+                break;
+            case 'HMAC-SHA1':
+                return 'stub data';
+                break;
+            default:
+                throw new ConfigException(ConfigException::UNKNOWN_SIGNATURE_METHOD, $this->signatureMethod);
+        }
+    }
+
+    protected function generateRSASHA1Signature(string $method, string $path, array $parameters)
+    {
+        if (!file_exists($this->privateKey)) {
+            throw new ConfigException(ConfigException::PRIVATE_KEY_NOT_FOUND, $this->privateKey);
+        }
+
+        $file = fopen($this->privateKey, 'r');
+        $contents = fread($file, 8192);
+        fclose($file);
+
+        $privateKey = openssl_pkey_get_private($contents);
+        if ($privateKey === false) {
+            throw new ConfigException(ConfigException::PRIVATE_KEY_INVALID, $this->privateKey);
+        }
+
+        $sbs = sprintf(
+            '%s&%s&%s',
+            $method,
+            $this->oauthEscape($this->endpoint . '/' . $path),
+            $this->oauthEscape($this->sortParameters($parameters))
+        );
+
+        openssl_sign($sbs, $signature, $privateKey);
+
+        openssl_free_key($privateKey);
+
+        return base64_encode($signature);
+    }
+
+    /**
+     * Sorts query parameters
+     * @param array $parameters
+     */
+    protected function sortParameters(array $parameters)
+    {
+        $elements = [];
+        ksort($parameters);
+        foreach ($parameters as $name => $value) {
+            if (is_array($value)) {
+                sort($value);
+                foreach ($value as $element) {
+                    array_push(
+                        $elements,
+                        sprintf('%s=%s', $this->oauthEscape($name), $this->oauthEscape($element))
+                    );
+                }
+                continue;
+            }
+            array_push(
+                $elements,
+                sprintf('%s=%s', $this->oauthEscape($name), $this->oauthEscape($value))
+            );
+        }
+        return join('&', $elements);
+    }
+
+    /**
+     * Escapes all special symbols for query
+     * @param string $string
+     */
+    protected function oauthEscape(string $string)
+    {
+        if (empty($string)) {
+            return '';
+        }
+        $string = rawurlencode($string);
+        $string = str_replace('+', '%20', $string);
+        $string = str_replace('!', '%21', $string);
+        $string = str_replace('*', '%2A', $string);
+        $string = str_replace('\'', '%27', $string);
+        $string = str_replace('(', '%28', $string);
+        $string = str_replace(')', '%29', $string);
```

Comment:
Is this the definitive list?

Should we not just urlencode everything?

What happens with characters like `<` and `{`?